skrub/_datetime_encoder.py

```diff
@@ -180,95 +170,145 @@ def fit(self, X: ArrayLike, y=None) -> "DatetimeEncoder":
+                    self.n_features_out_ += 1
+
+                if self.add_day_of_the_week:
+                    self.features_per_column_[col].append("day_of_week")
+                    self.n_features_out_ += 1
+
+    def transform(self, X, y=None):
+        """Transform `X` by replacing each datetime column with \
         corresponding numerical features.
 
         Parameters
         ----------
-        X : array-like, shape (``n_samples``, ``n_features``)
+        X : array-like of shape ``(n_samples, n_features)``
             The data to transform, where each column is a datetime feature.
         y : None
             Unused, only here for compatibility.
 
         Returns
         -------
-        ndarray, shape (``n_samples``, ``n_features_out_``)
+        X_out : ndarray of shape ``(n_samples, n_features_out_)``
             Transformed input.
         """
-        check_is_fitted(
-            self,
-            attributes=["n_features_in_", "n_features_out_", "features_per_column_"],
-        )
-        X = check_input(X)
-        if X.shape[1] != self.n_features_in_:
-            raise ValueError(
-                f"The number of features in the input data ({X.shape[1]}) "
-                "does not match the number of features "
-                f"seen during fit ({self.n_features_in_}). "
-            )
-        # Create a new array with the extracted features,
-        # choosing only features that weren't constant during fit
-        X_ = np.empty((X.shape[0], self.n_features_out_), dtype=np.float64)
-        idx = 0
-        for i in range(X.shape[1]):
-            for j, feature in enumerate(self.features_per_column_[i]):
-                X_[:, idx + j] = self._extract_from_date(X[:, i], feature)
-            idx += len(self.features_per_column_[i])
-        return X_
-
-    def get_feature_names_out(self, input_features=None) -> list[str]:
-        """Return clean feature names.
+        check_is_fitted(self)
+        self._check_n_features(X, reset=False)
+        self._check_feature_names(X, reset=False)
+        X = check_array(X, ensure_2d=True, force_all_finite=False, dtype=None)
+
+        return self._parse_datetime_cols(X)
+
+    def _parse_datetime_cols(self, X):
+        """Extract datetime features from the selected columns.
+
+        Parameters
+        ----------
+        X : ndarray of shape ``(n_samples, n_features)``
+
+        Returns
+        -------
+        X_out : ndarray of shape ``(n_samples, n_features_out_)``
+        """
+        columns = getattr(self, "feature_names_in_", list(range(X.shape[1])))
+        # X_out must be of dtype float64 to handle np.nan
+        X_out = np.empty((X.shape[0], self.n_features_out_), dtype=np.float64)
+        offset_idx = 0
+        for col_idx, col in enumerate(columns):
+            if col in self.features_per_column_:
+                # X_col is a DatetimeIndex
+                X_col = pd.to_datetime(X[:, col_idx], errors=self.errors)
```

Comment:
This also needs `format=self.format_per_column[col]`, after adding `_table_vectorizer._infer_date_format` logic.